src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,343 @@
+        K2INFO("Log stream does not create");
+        throw k2::dto::LogStreamExistError("Log stream does not create");
+    }
+    std::pair<String, uint32_t>& targetPlogInfo = writeToWAL ? _walInfo : _metaInfo;
+
+    if (targetPlogInfo.second + payload.getSize() < PLOG_MAX_SIZE){
+        return _client.append(targetPlogInfo.first, targetPlogInfo.second, std::move(payload)).
+        then([this, writeToWAL] (auto&& response){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()) {
+                throw k2::dto::PlogAppendError("unable to append the plog");
+            }
+            // Check weather there is a sealed request that did not receive the response
+            if (_switched && writeToWAL){
+                _requestWaiters.emplace_back(seastar::promise<>());
+                // if there is a flying sealed request, we should not notify the client until we receive the response of that sealed request
+                return _requestWaiters.back().get_future().
+                then([] (){
+                    return seastar::make_ready_future<>();
+                });
+            }
+            else{
+                return seastar::make_ready_future<>();
+            }
+        });
+    }
+    else{
+        // switch to a new plog
+        return _switchPlog(std::move(payload), writeToWAL);
+    }
+}
+
+
+seastar::future<> 
+LogStream::_switchPlog(Payload payload, bool writeToWAL){
+    std::pair<String, uint32_t>& targetPlogInfo = writeToWAL ? _walInfo : _metaInfo;
+
+    String new_plogId;
+    String sealed_plogId = targetPlogInfo.first;
+    uint32_t sealed_offest = targetPlogInfo.second;
+    
+    // change to the next available plog id
+    if (!writeToWAL){
+        if (_metadataPlogPool.size() == 0){
+            throw k2::dto::LogStreamBackupPlogError("no available backup metadata plog");
+        }
+        new_plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+        _metadataPlogPool.pop_back();
+    }
+    else{
+        if (_walPlogPool.size() == 0){
+            throw k2::dto::LogStreamBackupPlogError("no available backup wal plog");
+        }
+        new_plogId = _walPlogPool[_walPlogPool.size()-1];
+        _walPlogPool.pop_back();
+    }
+    targetPlogInfo.first = new_plogId;
+    targetPlogInfo.second = 0;
+    
+    _switched = true;
+    uint32_t current_offset = targetPlogInfo.second;
+    targetPlogInfo.second += payload.getSize();
+
+    return _client.append(targetPlogInfo.first, current_offset, std::move(payload))
+    .then([&, writeToWAL, sealed_plogId, new_plogId, sealed_offest] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw k2::dto::PlogAppendError("unable to append the plog");
+        }
+        // If sealed_offest == 0, then it means after the incoming writing operation, the current plog is not excceed the size limit
+        return _client.seal(sealed_plogId, sealed_offest)

```

Comment:
why seal after append new data, we can do it in parallel, right?

Also, what happens while the seal is ongoing, and append comes in and will it try to seal again (which is wrong)?

Reply:
Yes, I agree. We can do them as well as replenish the Plog pool in parallel.

 Seal the same Plog twice will not happen. If we find the old Plog needs to be sealed, we will switch to the backup Plog from the Plog pool immediately. Then when the new append comes, it will write to the new Plog instead of the one that needs to be sealed.